sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgBuilder.cs

```diff
@@ -0,0 +1,126 @@
+            instructionFactory = new UcfgInstructionFactory(semanticModel, objectFactory);
+        }
+
+        public UCFG Build(SyntaxNode syntaxNode, IMethodSymbol methodSymbol, IControlFlowGraph cfg)
+        {
+            var ucfg = objectFactory.CreateUcfg(syntaxNode, UcfgMethod.Create(methodSymbol));
+
+            ucfg.BasicBlocks.AddRange(cfg.Blocks.Select(CreateBasicBlock));
+
+            ucfg.Parameters.AddRange(methodSymbol.GetParameters().Select(p => p.Name));
+
+            if (syntaxNode is BaseMethodDeclarationSyntax methodDeclaration &&
+                TaintAnalysisEntryPointDetector.IsEntryPoint(methodSymbol))
+            {
+                var entryPointBlock = CreateEntryPointBlock(methodDeclaration, methodSymbol, blockIdProvider.Get(cfg.EntryBlock));
+                ucfg.BasicBlocks.Add(entryPointBlock);
+                ucfg.Entries.Add(entryPointBlock.Id);
+            }
+            else
+            {
+                ucfg.Entries.Add(blockIdProvider.Get(cfg.EntryBlock));
+            }
+
+            return ucfg;
+        }
+
+        private BasicBlock CreateBasicBlock(Block block)
+        {
+            var ucfgBlock = objectFactory.CreateUcfgBlock(blockIdProvider.Get(block));
+
+            ucfgBlock.Instructions.AddRange(
+                block.Instructions.Select(instructionFactory.Create).WhereNotNull());
+
+            if (block is JumpBlock jumpBlock &&
+                jumpBlock.JumpNode is ReturnStatementSyntax returnStatement)
+            {
+                ucfgBlock.Ret = objectFactory.CreateReturnExpression(returnStatement, returnStatement.Expression);
+            }
+
+            if (block is ExitBlock exit)
+            {
+                ucfgBlock.Ret = objectFactory.CreateReturnExpression();
+            }
+
+            if (ucfgBlock.TerminatorCase == BasicBlock.TerminatorOneofCase.None)
+            {
+                // No return was created from JumpBlock or ExitBlock, wire up the successor blocks
+                ucfgBlock.Jump = objectFactory.CreateJump(block.SuccessorBlocks.Select(blockIdProvider.Get).ToArray());
+            }
+
+            return ucfgBlock;
+        }
+
+        private BasicBlock CreateEntryPointBlock(BaseMethodDeclarationSyntax methodDeclaration,
+            IMethodSymbol methodSymbol, string currentEntryBlockId)
+        {
+            var ucfgBlock = objectFactory.CreateUcfgBlock(blockIdProvider.Get(new TemporaryBlock()));
+
+            ucfgBlock.Jump = objectFactory.CreateJump(currentEntryBlockId);
+
+            ucfgBlock.Instructions.Add(instructionFactory.Create(methodDeclaration));
+
+            foreach (var parameter in methodSymbol.Parameters)
+            {
+                ucfgBlock.Instructions.AddRange(CreateParameterInstructions(parameter));
+            }
+
+            return ucfgBlock;
+        }
+
+        private IEnumerable<Instruction> CreateParameterInstructions(IParameterSymbol parameter) =>
```

Comment:
I don't think this belongs here... You have a specialized class building instructions.